sql/2019/06_Fonts/06_09b.sql

```diff
@@ -0,0 +1,39 @@
+#standardSQL
+
+-- counts the number of pages given number of font-face
+
+CREATE TEMP FUNCTION parseCSS(stylesheet STRING)
+RETURNS INT64 LANGUAGE js AS '''
+   var css = parse(stylesheet, { silent: true })
+   var fontFace = {};
+
+   if (css.type === 'stylesheet' && css.stylesheet && css.stylesheet.rules) {
+     for (var rule of css.stylesheet.rules) {
+       if (rule.type === 'font-face' || rule.type === 'rule') {
+         if (rule.declarations && rule.declarations.length) {
+           for (var declaration of rule.declarations) {
+             if (declaration.type === 'declaration' && declaration.property === 'font-family') {
+               fontFace[declaration.value] = true
+             }
+           }
+         }
+       }
+     }
+   }
+
+   return Object.keys(fontFace).length;
+'''
+OPTIONS (library="gs://httparchive/lib/parse-css.js");
+
+SELECT
+  APPROX_QUANTILES(faces_per_page, 1000)[OFFSET(250)] AS p25_faces_per_page,
+  APPROX_QUANTILES(faces_per_page, 1000)[OFFSET(500)] AS median_faces_per_page,
+  APPROX_QUANTILES(faces_per_page, 1000)[OFFSET(750)] AS p75_faces_per_page
+FROM (
+  SELECT SUM(parseCSS(body)) as faces_per_page
+  FROM `httparchive.response_bodies.2017_01_01_*`
```

Comment:
I think joining response bodies with requests could be causing the timeout. I'll look into ways to optimize this.

FWIW changing to use the 1k tables runs the query successfully:

```sql
CREATE TEMP FUNCTION parseCSS(stylesheet STRING)
RETURNS INT64 LANGUAGE js AS '''
   var css = parse(stylesheet, { silent: true })
   var fontFace = {};
    if (css.type === 'stylesheet' && css.stylesheet && css.stylesheet.rules) {
     for (var rule of css.stylesheet.rules) {
       if (rule.type === 'font-face' || rule.type === 'rule') {
         if (rule.declarations && rule.declarations.length) {
           for (var declaration of rule.declarations) {
             if (declaration.type === 'declaration' && declaration.property === 'font-family') {
               fontFace[declaration.value] = true
             }
           }
         }
       }
     }
   }
    return Object.keys(fontFace).length;
'''
OPTIONS (library="gs://httparchive/lib/parse-css.js");

 SELECT
  APPROX_QUANTILES(faces_per_page, 1000)[OFFSET(250)] AS p25_faces_per_page,
  APPROX_QUANTILES(faces_per_page, 1000)[OFFSET(500)] AS median_faces_per_page,
  APPROX_QUANTILES(faces_per_page, 1000)[OFFSET(750)] AS p75_faces_per_page
FROM (
  SELECT SUM(parseCSS(body)) AS faces_per_page
  FROM `httparchive.almanac.response_bodies_*`
  JOIN (SELECT _TABLE_SUFFIX, url, pageid FROM `httparchive.almanac.summary_requests_*` WHERE type = 'css')
  USING (_TABLE_SUFFIX, url)
  GROUP BY pageid
)
```

So it seems to be a memory issue with joining 370M^2 rows.

Reply:
I'll try joining them I guess after I parse the css?

Reply:
Ok this has been haunting me all week but I think I've finally got a fix. See https://discuss.httparchive.org/t/analyzing-stylesheets-with-a-js-based-parser/1683/13?u=rviscomi

I've created a new table called `almanac.parsed_css` with the JSON output from the parser script for all ~80M stylesheets. It doesn't include inline styles, which was an oversight. I'll look into adding those to the table.

Here's an example of querying the parsed stylesheet JSON:

```sql
#standardSQL
CREATE TEMPORARY FUNCTION getAllValues(css STRING, property STRING)
RETURNS ARRAY<STRING> LANGUAGE js AS '''
try {
  var reduceValues = (values, rule) => {
    if (rule.type == 'media') {
      return reduceValues(values, rule.rules);
    }
    
    return values.concat(rule.declarations.filter(d => d.property == property).map(d => d.value));
  };
  var $ = JSON.parse(css);
  return $.stylesheet.rules.reduce(reduceValues, []);
} catch (e) {
  return [];
}
''';

SELECT
  APPROX_TOP_COUNT(color, 10)
FROM
  `httparchive.almanac.parsed_css`,
  UNNEST(getAllValues(css, 'color')) AS color
```